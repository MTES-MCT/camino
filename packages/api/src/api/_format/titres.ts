import { ITitre, ITitreDemarche } from '../../types.js'

import { entrepriseFormat } from './entreprises.js'
import { titreActiviteFormat } from './titres-activites.js'
import { titreDemarcheFormat } from './titres-demarches.js'
import { titreFormatFields } from './_fields.js'
import { AdministrationId } from 'camino-common/src/static/administrations.js'
import { isNullOrUndefined, onlyUnique } from 'camino-common/src/typescript-tools.js'
import { getGestionnairesByTitreTypeId } from 'camino-common/src/static/administrationsTitresTypes.js'
import { ACTIVITES_STATUTS_IDS } from 'camino-common/src/static/activitesStatuts.js'
import { FieldsTitre } from '../../database/queries/_options'

// optimisation possible pour un expert SQL
// remplacer le contenu de ce fichier
// par des requêtes SQL (dans /database/queries/titres)
// qui retournent les données directement formatées
export const titreFormat = (t: ITitre, fields: FieldsTitre = titreFormatFields) => {
  if ((t.confidentiel ?? false) === true) {
    // Si le titre est confidentiel, on a le droit de voir que son périmètre sur la carte
    t = {
      titreStatutId: t.titreStatutId,
      typeId: t.typeId,
      secteursMaritime: t.secteursMaritime,
      forets: t.forets,
      communes: t.communes,
      demarches: [] as ITitreDemarche[],
    } as ITitre
  }

  if (isNullOrUndefined(fields)) return t

  if (fields.demarches && t.demarches?.length) {
    t.demarches = t.demarches.map(td => titreDemarcheFormat(td, fields.demarches))
  }

  if (fields.surface && t.pointsEtape) {
    t.surface = t.pointsEtape.surface
  }

  if (fields.activites && t.activites?.length) {
    t.activites = t.activites.map(ta => {
      ta.titre = t

      return titreActiviteFormat(ta)
    })
  }

  if (t.activites?.length) {
    t.activitesAbsentes = t.activites.filter(({ activiteStatutId }) => activiteStatutId === ACTIVITES_STATUTS_IDS.ABSENT).length
    t.activitesEnConstruction = t.activites.filter(({ activiteStatutId }) => activiteStatutId === ACTIVITES_STATUTS_IDS.EN_CONSTRUCTION).length
  }

  if (fields.administrations) {
    t.administrations = titreAdministrationsGet(t)
  }

  t.titulaires = t.titulaires?.map(entrepriseFormat)

  t.amodiataires = t.amodiataires?.map(entrepriseFormat)

  return t
}

export const titreAdministrationsGet = (titre: ITitre): AdministrationId[] => {
  const ids: AdministrationId[] = getGestionnairesByTitreTypeId(titre.typeId)
    .filter(({ associee }) => !associee)
    .map(({ administrationId }) => administrationId)

  if (titre.administrationsLocales) {
    ids.push(...titre.administrationsLocales)
  }

  return ids.filter(onlyUnique)
}

export const titresFormat = (titres: ITitre[], fields = titreFormatFields) =>
  titres &&
  titres.reduce((acc: ITitre[], titre) => {
    const titreFormated = titreFormat(titre, fields)

    if (titreFormated) {
      acc.push(titreFormated)
    }

    return acc
  }, [])
