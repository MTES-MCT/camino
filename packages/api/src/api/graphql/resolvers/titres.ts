import { GraphQLResolveInfo } from 'graphql'

import { ITitre, ITitreColonneId, ITitreDemarche, IToken } from '../../../types'

import { debug } from '../../../config/index'
import {
  titreAdministrationsGet,
  titreFormat,
  titresFormat
} from '../../_format/titres'

import { fieldsBuild } from './_fields-build'

import {
  titreCreate,
  titreGet,
  titresCount,
  titresGet,
  titreUpsert,
  titreArchive
} from '../../../database/queries/titres'
import { userGet } from '../../../database/queries/utilisateurs'

import titreUpdateTask from '../../../business/titre-update'

import { titreUpdationValidate } from '../../../business/validations/titre-updation-validate'
import { domaineGet } from '../../../database/queries/metas'
import {
  canLinkTitres,
  getLinkConfig
} from 'camino-common/src/permissions/titres'
import { linkTitres } from '../../../database/queries/titres-titres'

const titre = async (
  { id }: { id: string },
  context: IToken,
  info: GraphQLResolveInfo
) => {
  try {
    const user = await userGet(context.user?.id)
    const fields = fieldsBuild(info)

    const titre = await titreGet(id, { fields, fetchHeritage: true }, user)

    if (!titre) return null

    return titreFormat(titre, fields)
  } catch (e) {
    if (debug) {
      console.error(e)
    }

    throw e
  }
}

const titres = async (
  {
    intervalle,
    page,
    colonne,
    ordre,
    ids,
    perimetre,
    typesIds,
    domainesIds,
    statutsIds,
    substancesLegalesIds,
    entreprisesIds,
    substances,
    noms,
    entreprises,
    references,
    territoires,
    demandeEnCours
  }: {
    intervalle?: number | null
    page?: number | null
    colonne?: ITitreColonneId | null
    ordre?: 'asc' | 'desc' | null
    ids: string[]
    perimetre?: number[] | null
    typesIds: string[]
    domainesIds: string[]
    statutsIds: string[]
    substancesLegalesIds: string[]
    entreprisesIds: string[]
    substances: string
    noms: string
    entreprises: string
    references: string
    territoires: string
    demandeEnCours: boolean | null
  },
  context: IToken,
  info: GraphQLResolveInfo
) => {
  try {
    const user = await userGet(context.user?.id)
    const fields = fieldsBuild(info).elements

    const [titres, total] = await Promise.all([
      titresGet(
        {
          intervalle,
          page,
          colonne,
          ordre,
          ids,
          perimetre,
          typesIds,
          domainesIds,
          statutsIds,
          substancesLegalesIds,
          entreprisesIds,
          substances,
          noms,
          entreprises,
          references,
          territoires,
          demandeEnCours
        },
        { fields },
        user
      ),
      titresCount(
        {
          ids,
          typesIds,
          domainesIds,
          statutsIds,
          substancesLegalesIds,
          entreprisesIds,
          substances,
          noms,
          entreprises,
          references,
          territoires,
          demandeEnCours
        },
        { fields: {} },
        user
      )
    ])

    const titresFormatted = titres && titresFormat(titres, fields)

    return {
      elements: titresFormatted,
      page,
      intervalle,
      ordre,
      colonne,
      total
    }
  } catch (e) {
    if (debug) {
      console.error(e)
    }

    throw e
  }
}

/**
 * TODO 2022-07-12 enlever cette fonction et nettoyer l'ui
 * @deprecated Not used by frontend, titreDemandeCreer is used instead
 */
const titreCreer = async (
  { titre }: { titre: ITitre },
  context: IToken,
  info: GraphQLResolveInfo
) => {
  try {
    const user = await userGet(context.user?.id)

    const domaine = await domaineGet(
      titre.domaineId,
      { fields: { titresTypes: { id: {} } } },
      user
    )
    const titreType = domaine?.titresTypes.find(tt => tt.id === titre.typeId)

    if (!user || !titreType || !titreType.titresCreation)
      throw new Error('droits insuffisants')

    // insert le titre dans la base
    titre = await titreCreate(titre, { fields: {} })

    await titreUpdateTask(titre.id)

    const fields = fieldsBuild(info)

    const titreUpdated = await titreGet(titre.id, { fields }, user)

    return titreUpdated && titreFormat(titreUpdated)
  } catch (e) {
    if (debug) {
      console.error(e)
    }

    throw e
  }
}

// TODO 2022-07-19 à mettre en REST mais il faut documenter les routes
export const titreLiaisonsModifier = async (
  { titreId, titreFromIds }: { titreId: string; titreFromIds: string[] },
  context: IToken
): Promise<boolean> => {
  try {
    const user = await userGet(context.user?.id)

    const titre = await titreGet(
      titreId,
      {
        fields: {
          administrationsGestionnaires: { id: {} },
          administrationsLocales: { id: {} },
          demarches: { id: {} }
        }
      },
      user
    )

    if (!titre) throw new Error("le titre n'existe pas")

    if (!titre.administrationsGestionnaires || !titre.administrationsLocales) {
      throw new Error('les administrations ne sont pas chargées')
    }

    const administrations = titreAdministrationsGet(titre)
    if (
      !canLinkTitres(
        user,
        administrations.map(({ id }) => id)
      )
    )
      throw new Error('droits insuffisants')

    if (!titre.demarches) {
      throw new Error('les démarches ne sont pas chargées')
    }

    const titresFrom = await titresGet(
      { ids: titreFromIds },
      { fields: { id: {} } },
      user
    )

    checkTitreLinks(titre, titreFromIds, titresFrom, titre.demarches)

    await linkTitres({ linkTo: titreId, linkFrom: titreFromIds })

    return true
  } catch (e) {
    if (debug) {
      console.error(e)
    }

    throw e
  }
}

export const checkTitreLinks = (
  titre: Pick<ITitre, 'typeId'>,
  titreFromIds: string[],
  titresFrom: ITitre[],
  demarches: ITitreDemarche[]
) => {
  const linkConfig = getLinkConfig(titre.typeId, demarches)
  if (!linkConfig) {
    throw new Error('ce titre ne peut pas être lié à d’autres titres')
  }

  if (linkConfig.count === 'single' && titreFromIds.length > 1) {
    throw new Error('ce titre peut avoir un seul titre lié')
  }

  if (titresFrom.length !== titreFromIds.length) {
    throw new Error('droit insuffisant')
  }

  if (linkConfig) {
    if (titresFrom.some(({ typeId }) => typeId !== linkConfig.typeId)) {
      throw new Error(
        `un titre de type ${titre.typeId} ne peut-être lié qu’à un titre de type ${linkConfig.typeId}`
      )
    }
  }
}

const titreModifier = async (
  { titre }: { titre: ITitre },
  context: IToken,
  info: GraphQLResolveInfo
) => {
  try {
    const user = await userGet(context.user?.id)

    const titreOld = await titreGet(
      titre.id,
      { fields: { titresAdministrations: { id: {} } } },
      user
    )

    if (!titreOld) throw new Error("le titre n'existe pas")

    if (!titreOld.modification) throw new Error('droits insuffisants')

    const rulesErrors = await titreUpdationValidate(titre, titreOld, user)

    if (rulesErrors.length) {
      throw new Error(rulesErrors.join(', '))
    }

    const fields = fieldsBuild(info)

    // on doit utiliser upsert (plutôt qu'un simple update)
    // car le titre contient des références (tableau d'objet)
    await titreUpsert(titre, { fields })

    await titreUpdateTask(titre.id)

    const titreUpdated = await titreGet(titre.id, { fields }, user)

    return titreUpdated && titreFormat(titreUpdated)
  } catch (e) {
    if (debug) {
      console.error(e)
    }

    throw e
  }
}

const titreSupprimer = async ({ id }: { id: string }, context: IToken) => {
  const user = await userGet(context.user?.id)

  const titreOld = await titreGet(
    id,
    {
      fields: {
        demarches: { etapes: { id: {} } },
        activites: { id: {} }
      }
    },
    user
  )

  if (!titreOld) throw new Error("le titre n'existe pas")

  if (!titreOld.suppression) throw new Error('droits insuffisants')

  await titreArchive(id)

  return titreOld.slug
}

export { titre, titres, titreCreer, titreModifier, titreSupprimer }
