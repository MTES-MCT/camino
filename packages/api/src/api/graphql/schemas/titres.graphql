# import * from 'scalars.graphql'
# import * from 'administrations.graphql'
# import * from 'entreprises.graphql'
# import * from 'titres-demarches.graphql'
# import * from 'metas.graphql'
# import * from 'substances.graphql'
# import * from 'points.graphql'
# import * from 'geojsons.graphql'
# import * from 'points.graphql'
# import * from 'territoires.graphql'
# import * from 'titres-activites.graphql'

type Titres {
  elements: [Titre]
  page: Int
  intervalle: Int
  ordre: String
  colonne: String
  total: Int
}

"Titre minier"
type Titre {
  """
  Id unique
  """
  id: ID

  """
  Slug composé de la concaténation:
  - du type en 3 lettres
  - du slug du nom du titre
  - de l'année d'octroi
  """
  slug: String

  "Nom du titre"
  nom: String

  "Type du titre"
  type: TitreType!

  "Domaine minier auquel se rattache le titre"
  domaine: Domaine!

  "DomaineId minier auquel se rattache le titre"
  domaineId: String

  "Références métier du titre"
  references: [Reference]

  "Statut du titre"
  titreStatutId: String!

  "Substances concernées par le titre dans son dernier état de validité"
  substances: [String!]

  "Coordonnées x,y du titre"
  coordonnees: Coordonnees

  "Coordonnées des points du périmètre géographique du titre dans son dernier état de validité"
  points: [Point]

  "Geojson du contour du périmètre géographique du titre dans son dernier état de validité"
  geojsonMultiPolygon: GeojsonMultiPolygon

  "Geojson de la liste de points du périmètre géographique du titre dans son dernier état de validité"
  geojsonPoints: GeojsonPoints

  "Point représentant le centre géographique du titre"
  geojsonCentre: GeojsonCentre

  "Titualires du titre dans son dernier état de validité"
  titulaires: [Entreprise]

  "Amodiataires du titre dans son dernier état de validité"
  amodiataires: [Entreprise]

  "Administrations en charge du titre dans son dernier état de validité"
  administrations: [String]

  "Territoires du titre dans son dernier état de validité"
  communes: [Commune]

  "Date de début d'un titre"
  dateDebut: String

  "Date de fin d'un titre"
  dateFin: String

  "Date de demande initiale d'un titre"
  dateDemande: String

  "Démarches effectuées sur le titre"
  demarches: [Demarche]

  "Surface en Km² du titre dans son dernier état de validité"
  surface: Float

  "Propriétés spécifiques au titre"
  contenu: Json

  "Activités"
  activites: [Activite]

  "Nombre d'activités déposées"
  activitesDeposees: Int

  "Nombre d'activités en construction"
  activitesEnConstruction: Int

  "Nombre d'activités absentes"
  activitesAbsentes: Int

  "Forets"
  forets: [Foret]

  "Zones du SDOM de Guyane"
  sdomZones: [SDOMZone]

  "Secteurs Maritime"
  secteursMaritime: [String]

  modification: Boolean
  suppression: Boolean
  demarchesCreation: Boolean

  travauxCreation: Boolean
  doublonTitre: Titre

  "L’utilisateur est abonné au titre"
  abonnement: Boolean
}

"Références métier d'un titre minier"
type Reference {
  "Type de références"
  referenceTypeId: String!

  "Nom de la référence"
  nom: String!
}

"Titre minier"
input InputTitreCreation {
  nom: String!
  typeId: ID!
  domaineId: ID!
  references: [InputReference]
}

"Titre minier"
input InputTitreModification {
  """
  Id composée de la concaténation:
  - du type en 3 lettres
  - du slug du nom du titre
  - de l'année d'octroi
  """
  id: ID!
  nom: String!
  references: [InputReference]
}

input InputTitreLiaisonsModification {
  id: ID!
  titreFromIds: [ID]!
}

input InputId {
  id: ID
}

input InputReference {
  referenceTypeId: String!
  nom: String!
}
