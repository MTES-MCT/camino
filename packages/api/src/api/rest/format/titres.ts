import { ITitre, Index } from '../../../types.js'
import { SubstancesLegale } from 'camino-common/src/static/substancesLegales.js'
import { isNotNullNorUndefined } from 'camino-common/src/typescript-tools.js'
import { TitresStatuts } from 'camino-common/src/static/titresStatuts.js'
import { ReferencesTypes } from 'camino-common/src/static/referencesTypes.js'
import { getDomaineId, getTitreTypeType } from 'camino-common/src/static/titresTypes.js'
import { TitresTypesTypes } from 'camino-common/src/static/titresTypesTypes.js'
import { Domaines } from 'camino-common/src/static/domaines.js'
import { getFacadesComputed, SecteursMaritimes } from 'camino-common/src/static/facades.js'
import { territoiresFind } from 'camino-common/src/territoires.js'
import { Administrations } from 'camino-common/src/static/administrations.js'
import { valeurFind } from 'camino-common/src/sections.js'
import { titreSectionsGet } from '../titre-contenu.js'
import { CaminoDate } from 'camino-common/src/date.js'
import { titreDemarcheSortAsc } from '../../../business/utils/titre-elements-sort-asc.js'
import { titreDateDemandeFind } from '../../../business/rules/titre-date-demande-find.js'
import { Forets } from 'camino-common/src/static/forets.js'
import { Pool } from 'pg'
import { CommuneId } from 'camino-common/src/static/communes.js'
import { getCommunesIndex } from '../../../database/queries/communes.js'

const getFacadesMaritimeCell = (secteursMaritime: SecteursMaritimes[], separator: string): string =>
  getFacadesComputed(secteursMaritime)
    .map(({ facade, secteurs }) => `${facade} (${secteurs.join(', ')})`)
    .join(separator)

const titreContenuTableFormat = (titre: ITitre): Record<string, string> => {
  const sections = titreSectionsGet(titre)

  return sections.reduce((result, section) => {
    const subResult = section.elements.reduce<Record<string, string>>((acc, element) => {
      acc[element.nom ?? element.id] = valeurFind(element)

      return acc
    }, {})

    return { ...result, ...subResult }
  }, {})
}

export const titresTableFormat = async (pool: Pool, titres: ITitre[]) => {
  const communesIndex = await getCommunesIndex(
    pool,
    titres.flatMap(titre => titre.communes?.map(({ id }) => id) ?? [])
  )

  return titres.map(titre => {
    if (!titre.secteursMaritime) {
      throw new Error('les secteurs maritimes ne sont pas chargés')
    }

    const { communes, departements, regions } = territoiresFind(communesIndex, titre.communes, titre.secteursMaritime)

    const titreReferences = titre.references
      ? titre.references.reduce((titreReferences: Index<string>, reference) => {
          titreReferences[`reference_${ReferencesTypes[reference.referenceTypeId].nom}`] = reference.nom

          return titreReferences
        }, {})
      : {}

    const separator = ';'

    const { dateDebut, dateFin, dateDemande } = getTitreDates(titre)

    const titreNew = {
      id: titre.slug,
      nom: titre.nom,
      type: TitresTypesTypes[getTitreTypeType(titre.typeId)].nom,
      domaine: Domaines[getDomaineId(titre.typeId)].nom,
      date_debut: dateDebut,
      date_fin: dateFin,
      date_demande: dateDemande,
      statut: isNotNullNorUndefined(titre.titreStatutId) ? TitresStatuts[titre.titreStatutId].nom : '',
      substances: titre.substances?.map(substanceId => SubstancesLegale[substanceId].nom)?.join(separator),
      'surface renseignee km2': titre.surface,
      'communes (surface calculee km2)': communes.map(({ nom, surface }) => `${nom} ${surface !== null ? ` (${surface})` : ''}`).join(separator),
      forets: titre.forets?.map(fId => Forets[fId].nom).join(separator),
      facades_maritimes: getFacadesMaritimeCell(titre.secteursMaritime, separator),
      departements: departements.join(separator),
      regions: regions.join(separator),
      administrations_noms: titre.administrations!.map(id => Administrations[id].nom).join(separator),
      titulaires_noms: titre.titulaires?.map(e => e.nom).join(separator),
      titulaires_adresses: titre.titulaires?.map(e => `${e.adresse} ${e.codePostal} ${e.commune}`).join(separator),
      titulaires_legal: titre.titulaires?.map(e => e.legalEtranger || e.legalSiren).join(separator),
      titulaires_categorie: titre.titulaires?.map(e => e.categorie).join(separator),
      amodiataires_noms: titre.amodiataires?.map(e => e.nom).join(separator),
      amodiataires_adresses: titre.amodiataires?.map(e => `${e.adresse} ${e.codePostal} ${e.commune}`).join(separator),
      amodiataires_legal: titre.amodiataires?.map(e => e.legalEtranger || e.legalSiren).join(separator),
      amodiataires_categorie: titre.amodiataires?.map(e => e.categorie).join(separator),
      geojson: JSON.stringify(titre.geojsonMultiPolygon),
      ...titreReferences,
      ...titreContenuTableFormat(titre),
    }

    return titreNew
  })
}

const titreGeojsonPropertiesFormat = (communesIndex: Record<CommuneId, string>, titre: ITitre) => {
  if (!titre.secteursMaritime) {
    throw new Error('les secteurs maritimes ne sont pas chargés')
  }
  const { departements, regions } = territoiresFind(communesIndex, titre.communes, titre.secteursMaritime)

  const { dateDebut, dateFin, dateDemande } = getTitreDates(titre)

  return {
    id: titre.slug,
    nom: titre.nom,
    type: TitresTypesTypes[getTitreTypeType(titre.typeId)].nom,
    domaine: Domaines[getDomaineId(titre.typeId)].nom,
    date_debut: dateDebut,
    date_fin: dateFin,
    date_demande: dateDemande,
    statut: isNotNullNorUndefined(titre.titreStatutId) ? TitresStatuts[titre.titreStatutId].nom : '',
    substances: titre.substances?.map(substanceId => SubstancesLegale[substanceId].nom) || null,
    surface_totale: titre.surface,
    communes: titre.communes?.map(({ id }) => communesIndex[id]),
    surface_par_communes: titre.communes?.map(({ id, surface }) => ({ nom: communesIndex[id], surface: Math.round((surface ?? 0) / 100) / 10000 })),
    forets: titre.forets?.map(fId => Forets[fId].nom),
    facades_maritimes: getFacadesComputed(titre.secteursMaritime).map(({ facade }) => facade),
    departements,
    regions,
    administrations_noms: titre.administrations!.map(id => Administrations[id].nom),
    titulaires_noms: titre.titulaires?.map(e => e.nom) || null,
    titulaires_legal: titre.titulaires?.map(e => e.legalEtranger || e.legalSiren) || null,
    amodiataires_noms: titre.amodiataires?.map(e => e.nom) || null,
    amodiataires_legal: titre.amodiataires?.map(e => e.legalEtranger || e.legalSiren) || null,
    references: titre.references && titre.references.map(reference => `${ReferencesTypes[reference.referenceTypeId].nom}: ${reference.nom}`),
    ...titreContenuTableFormat(titre),
  }
}

const getTitreDates = (titre: Pick<ITitre, 'demarches'>): { dateDebut: CaminoDate | null; dateFin: CaminoDate | null; dateDemande: CaminoDate | null } => {
  const sortedDemarches = titreDemarcheSortAsc(titre.demarches ?? [])

  return {
    dateDebut: sortedDemarches.find(demarche => !!demarche.demarcheDateDebut)?.demarcheDateDebut ?? null,
    dateFin: sortedDemarches.reverse().find(demarche => demarche.demarcheDateDebut)?.demarcheDateFin ?? null,
    dateDemande: titreDateDemandeFind(sortedDemarches),
  }
}

export const titreGeojsonFormat = async (pool: Pool, titre: ITitre) => {
  const communesIndex = await getCommunesIndex(pool, titre.communes?.map(({ id }) => id) ?? [])

  return {
    type: 'FeatureCollection',
    properties: titreGeojsonPropertiesFormat(communesIndex, titre),
    features: titre.geojsonPoints ? [titre.geojsonMultiPolygon].concat(titre.geojsonPoints.features) : titre.geojsonMultiPolygon,
  }
}

export const titresGeojsonFormat = async (pool: Pool, titres: ITitre[]) => {
  const communesIndex = await getCommunesIndex(
    pool,
    titres.flatMap(titre => titre.communes?.map(({ id }) => id) ?? [])
  )

  return {
    type: 'FeatureCollection',
    features: titres.map(titre => ({
      type: 'Feature',
      geometry: titre.geojsonMultiPolygon?.geometry,
      properties: titreGeojsonPropertiesFormat(communesIndex, titre),
    })),
  }
}
