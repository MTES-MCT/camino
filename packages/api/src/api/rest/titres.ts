import { titresGet } from '../../database/queries/titres'

import { userGet } from '../../database/queries/utilisateurs'

import {
  ADMINISTRATION_IDS,
  AdministrationId
} from 'camino-common/src/administrations'
import express from 'express'
import { constants } from 'http2'
import { DOMAINES_IDS } from 'camino-common/src/domaines'
import { TITRES_TYPES_TYPES_IDS } from 'camino-common/src/titresTypesTypes'
import { ITitre, IUser, ITitreReference, ITitreDemarche } from '../../types'
import {
  CommonTitreDREAL,
  CommonTitreONF,
  CommonTitrePTMG
} from 'camino-common/src/titres'
import {
  toMachineEtapes,
  whoIsBlocking
} from '../../business/rules-demarches/machine-helper'
import {
  demarcheDefinitionFind,
  isDemarcheDefinitionMachine
} from '../../business/rules-demarches/definitions'
import { CustomResponse } from './express-type'
import { userSuper } from '../../database/user-super'
import Utilisateurs from '../../database/models/utilisateurs'
import { NotNullableKeys } from 'camino-common/src/typescript-tools'

type MyTitreRef = { type: NonNullable<ITitreReference['type']> } & Omit<
  ITitreReference,
  'type'
>

export const titresONF = async (
  req: express.Request,
  res: CustomResponse<CommonTitreONF[]>
) => {
  const userId = (req.user as unknown as IUser | undefined)?.id

  const user = await userGet(userId)

  const onf = ADMINISTRATION_IDS['OFFICE NATIONAL DES FORÊTS']

  if (!user?.administrations?.some(({ id }) => id === onf)) {
    res.sendStatus(constants.HTTP_STATUS_FORBIDDEN)
  } else {
    const titresAvecOctroiArm = await titresArmAvecOctroi(user, onf)
    res.json(
      titresAvecOctroiArm.map(({ titre, references, octARM, blockedByMe }) => {
        const dateCompletudePTMG =
          octARM.etapes.find(etape => etape.typeId === 'mcp')?.date || ''

        const dateReceptionONF =
          octARM.etapes.find(etape => etape.typeId === 'mcr')?.date || ''

        const dateCARM =
          octARM.etapes.find(etape => etape.typeId === 'sca')?.date || ''

        return {
          id: titre.id,
          slug: titre.slug,
          nom: titre.nom,
          statut: titre.statut,
          references,
          titulaires: titre.titulaires,
          dateCompletudePTMG,
          dateReceptionONF,
          dateCARM,
          enAttenteDeONF: blockedByMe
        }
      })
    )
  }
}

type TitreSanitize = NotNullableKeys<
  Required<Pick<ITitre, 'slug' | 'titulaires' | 'statut'>>
> &
  Pick<ITitre, 'typeId' | 'id' | 'nom'>
type TitreDemarcheSanitize = NotNullableKeys<
  Required<Pick<ITitreDemarche, 'etapes' | 'typeId'>>
>

type TitreArmAvecOctroi = {
  titre: TitreSanitize
  references: MyTitreRef[]
  octARM: TitreDemarcheSanitize
  blockedByMe: boolean
}

async function titresArmAvecOctroi(
  user: null | Utilisateurs | undefined,
  administrationId: AdministrationId
) {
  const filters = {
    domainesIds: [DOMAINES_IDS.METAUX],
    typesIds: [TITRES_TYPES_TYPES_IDS.AUTORISATION_DE_RECHERCHE],
    statutsIds: ['dmi', 'mod', 'val']
  }
  const titresAutorises = await titresGet(
    filters,
    {
      fields: { id: {} }
    },
    user
  )
  const titresAutorisesIds = titresAutorises.map(({ id }) => id)
  const titres = await titresGet(
    { ...filters, ids: titresAutorisesIds, colonne: 'nom' },
    {
      fields: {
        statut: { id: {} },
        references: { type: { id: {} } },
        titulaires: { id: {} },
        demarches: { etapes: { id: {} } }
      }
    },
    userSuper
  )
  const titresAvecOctroiArm: TitreArmAvecOctroi[] = titres
    .map<TitreArmAvecOctroi | null>((titre: ITitre) => {
      if (titre.slug === undefined) {
        return null
      }
      if (!titre.statut) {
        throw new Error('le statut du titre n’est pas chargé')
      }

      if (!titre.references) {
        throw new Error('les références ne sont pas chargées')
      }

      const references = titre.references.filter(
        (reference: ITitreReference): reference is MyTitreRef =>
          !!reference.type && !!reference.type.nom && !!reference.nom
      )
      if (titre.references.length !== references.length) {
        throw new Error('le type de référence n’est pas chargé')
      }

      if (!titre.titulaires) {
        throw new Error('les titulaires ne sont pas chargés')
      }

      if (!titre.demarches) {
        throw new Error('les démarches ne sont pas chargées')
      }

      const octARM = titre.demarches.find(demarche => demarche.typeId === 'oct')

      if (!octARM) {
        return null
      }

      if (!octARM.etapes) {
        throw new Error('les étapes ne sont pas chargées')
      }

      const hasMachine = isDemarcheDefinitionMachine(
        demarcheDefinitionFind(titre.typeId, octARM.typeId, octARM.etapes)
      )
      const blockedByMe: boolean =
        hasMachine &&
        whoIsBlocking(toMachineEtapes(octARM.etapes)).includes(administrationId)

      // TODO 2022-06-08 wait for typescript to get better at type interpolation
      return {
        titre: titre as TitreSanitize,
        references,
        octARM: octARM as TitreDemarcheSanitize,
        blockedByMe
      }
    })
    .filter(
      (titre: TitreArmAvecOctroi | null): titre is TitreArmAvecOctroi =>
        titre !== null
    )

  return titresAvecOctroiArm
}

export const titresPTMG = async (
  req: express.Request,
  res: CustomResponse<CommonTitrePTMG[]>
) => {
  const userId = (req.user as unknown as IUser | undefined)?.id

  const user = await userGet(userId)

  const administrationId = ADMINISTRATION_IDS['PÔLE TECHNIQUE MINIER DE GUYANE']

  if (!user?.administrations?.some(({ id }) => id === administrationId)) {
    res.sendStatus(constants.HTTP_STATUS_FORBIDDEN)
  } else {
    const titresFormated: CommonTitrePTMG[] = (
      await titresArmAvecOctroi(user, administrationId)
    ).map(({ titre, references, blockedByMe }) => {
      return {
        id: titre.id,
        slug: titre.slug,
        nom: titre.nom,
        statut: titre.statut,
        references,
        titulaires: titre.titulaires,
        enAttenteDePTMG: blockedByMe
      }
    })

    res.json(titresFormated)
  }
}

type DrealTitreSanitize = NotNullableKeys<
  Required<Pick<ITitre, 'slug' | 'titulaires' | 'statut' | 'type'>>
> &
  Pick<
    ITitre,
    | 'typeId'
    | 'id'
    | 'nom'
    | 'domaineId'
    | 'activitesEnConstruction'
    | 'activitesAbsentes'
  >

interface TitreDrealAvecReferences {
  titre: DrealTitreSanitize
  references: MyTitreRef[]
}
export const titresDREAL = async (
  req: express.Request,
  res: CustomResponse<CommonTitreDREAL[]>
) => {
  const userId = (req.user as unknown as IUser | undefined)?.id

  const user = await userGet(userId)

  const filters = {
    statutsIds: ['dmi', 'mod', 'val']
  }

  const titres = await titresGet(
    { ...filters, colonne: 'nom' },
    {
      fields: {
        statut: { id: {} },
        type: { id: {} },
        references: { type: { id: {} } },
        titulaires: { id: {} },
        activites: { id: {} }
      }
    },
    user
  )

  const titresFormated: CommonTitreDREAL[] = titres
    .map((titre: ITitre): TitreDrealAvecReferences | null => {
      if (titre.slug === undefined) {
        return null
      }
      if (!titre.statut) {
        throw new Error('le statut du titre n’est pas chargé')
      }

      if (!titre.type) {
        throw new Error('les types de titres ne sont pas chargées')
      }

      if (!titre.titulaires) {
        throw new Error('les titulaires ne sont pas chargés')
      }

      if (!titre.references) {
        throw new Error('les références ne sont pas chargées')
      }

      if (!titre.activites) {
        throw new Error('les activités ne sont pas chargées')
      }

      const references = titre.references.filter(
        (reference: ITitreReference): reference is MyTitreRef =>
          !!reference.type && !!reference.type.nom && !!reference.nom
      )
      if (titre.references.length !== references.length) {
        throw new Error('le type de référence n’est pas chargé')
      }

      return { titre: titre as DrealTitreSanitize, references }
    })
    .filter(
      (
        titre: TitreDrealAvecReferences | null
      ): titre is TitreDrealAvecReferences => titre !== null
    )
    .map(({ titre, references }) => {
      return {
        id: titre.id,
        slug: titre.slug,
        nom: titre.nom,
        statut: titre.statut,
        typeId: titre.type.typeId,
        references,
        domaineId: titre.domaineId,
        titulaires: titre.titulaires,
        // pour une raison inconnue les chiffres sortent parfois en tant que string...., par exemple pour les titres
        activitesEnConstruction:
          typeof titre.activitesEnConstruction === 'string'
            ? parseInt(titre.activitesEnConstruction, 10)
            : titre.activitesEnConstruction ?? 0,
        activitesAbsentes:
          typeof titre.activitesAbsentes === 'string'
            ? parseInt(titre.activitesAbsentes, 10)
            : titre.activitesAbsentes ?? 0
      }
    })

  res.json(titresFormated)
}
