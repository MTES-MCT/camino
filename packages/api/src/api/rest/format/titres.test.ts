import { titreTerritoiresFind } from './titres'
import { DEPARTEMENT_IDS } from 'camino-common/src/static/departement'

describe('titreTerritoiresFind', () => {
  test('titreTerritoiresFind', () => {
    expect(
      titreTerritoiresFind([
        { nom: 'Fl√©e', surface: 100000, departementId: DEPARTEMENT_IDS.Sarthe },
        {
          nom: 'Montval-sur-loir',
          surface: 105020,
          departementId: DEPARTEMENT_IDS.Sarthe
        },
        {
          nom: 'Tours',
          surface: 99999,
          departementId: DEPARTEMENT_IDS['Indre-et-Loire']
        }
      ])
    ).toMatchSnapshot()
  })
})
