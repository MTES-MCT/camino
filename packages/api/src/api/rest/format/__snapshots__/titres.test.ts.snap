// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`titreTerritoiresFind titreTerritoiresFind 1`] = `
Object {
  "communes": Array [
    "Fl√©e (0.1)",
    "Montval-sur-loir (0.105)",
    "Tours (0.1)",
  ],
  "departements": Array [
    "Sarthe",
    "Indre-et-Loire",
  ],
  "regions": Array [
    "Pays de la Loire",
    "Centre-Val de Loire",
  ],
}
`;
